synapse/config/database.py

```diff
@@ -12,12 +12,77 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import logging
 import os
 from textwrap import indent
+from typing import List
 
 import yaml
 
-from ._base import Config
+from twisted.enterprise import adbapi
+
+from synapse.config._base import Config, ConfigError
+from synapse.storage.engines import create_engine
```

Comment:
having `config` import from `synapse.storage` sounds like a great way to end up with horrible import loops (I found a fun one yesterday where if you `import synapse.storage` first you're ok, but if you `import synapse.state` it all explodes).

It seems to me that `DatabaseConnectionConfig` is doing more than a config object should: why can't the call to create_engine, and the `get_pool` and `make_conn` and co move into the `Database` class?

An alternative looks like it might be to move `create_engine` into `config`, but that feels wrong to me.

Reply:
(that said: a big +1 to moving these things out of the HomeServer god object)

Reply:
I don't disagree, it is a pain. At the moment the `Database` object represents a fully prepared and configured database, we could move the preparation into the `__init__`, but then we need want to reuse the connection to give to the data stores, which is fiddly if we create the connection in the `Database` :(

Reply:
> then we need want to reuse the connection to give to the data stores, which is fiddly if we create the connection in the Database :(

I'm not quite following this?

Reply:
Basically its a bit hard to see how to rewrite https://github.com/matrix-org/synapse/pull/6513/files#diff-23406f76811c2c98a6ed1758c0bda2abR42-R54 if we move `make_conn` to `Database`

Reply:
I see. Make it a top-level function, in `synapse.storage.database` (or similar), which takes a `DatabaseConfig` as a param?

Reply:
Hmm, ok, I can give that a go

Reply:
So https://github.com/matrix-org/synapse/pull/6513/commits/439e043890a3a8ed365a8b3d5db5ab75fba3dd5c is my attempt at that, the tests are a bit hacky but that's true of them already.